src/OrchardCore.Modules/OrchardCore.Queries/Deployment/QueryBasedContentDeploymentSource.cs

```diff
@@ -18,28 +18,33 @@ public QueryBasedContentDeploymentSource(IQueryManager queryManager)
 
         public async Task ProcessDeploymentStepAsync(DeploymentStep step, DeploymentPlanResult result)
         {
-            var contentStep = step as QueryBasedContentDeploymentStep;
+            var queryDeploymentStep = step as QueryBasedContentDeploymentStep;
 
-            if (contentStep == null)
+            if (queryDeploymentStep == null)
             {
                 return;
             }
 
             var data = new JArray();
 
-            var query = await _queryManager.GetQueryAsync(contentStep.QueryName) as SqlQuery;
+            dynamic query = await _queryManager.GetQueryAsync(queryDeploymentStep.QueryName);
```

Comment:
Let's see if we can find a better solution than using `dynamic` and then tying the implementation to the Lucene and SQL Query types here and in the driver.

I don't have a better suggestion that to try to run the query and see what it returns, but it might not return anything by design. Perhaps we need to refactor `Query` and add a property to signal what `ReturnContentItems` and `ReturnDocuments` does (though these two are not the same indeed).

Reply:
Perhaps @sebastienros you have a better idea, since `Query` was originally written by you?

Reply:
> I don't have a better suggestion that to try to run the query and see what it returns, but it might not return anything by design. Perhaps we need to refactor `Query` and add a property to signal what `ReturnContentItems` and `ReturnDocuments` does (though these two are not the same indeed).

because we have this warning message, saying that user needs to make sure that the Query is returning content items, we could also remove those additional checks and avoid using dynamic in those two places

another idea, is to perform some additional checks inside **QueryBasedContentDeploymentSource.cs** when there is at least one item returned by the Query, and check if this item looks like **ContentItem** type and only add it if condition is met

![image](https://user-images.githubusercontent.com/6403130/182653015-19165000-8faa-4436-9c39-8508cfd3d9ea.png)

I'm waiting for Sebastien's thoughts.

Reply:
The driver, unfortunately, has the same issue, so we need to figure out a way to get rid of this tight coupling altogether.

Yeah, that checking the return object idea is what I mentioned above and I don't think is good.

Reply:
I understand that Query can be inherited by SqlQuery and LuceneQuery and ElasticsearchQuery. But here you want to export content items based on SQL Queries so why not using : 

```
var query = await _queryManager.GetQueryAsync(contentStep.QueryName)

if (typeof(SqlQuery) == query.GetType())
{
   //do something
}
```

And if the casted query doesn't contain any

Reply:
I've just changed PR title, because I see why you are confused. When I've opened this PR, I thought this will be only used with SQL Queries, but then Piedone told me it should work with any type of query which still returns content items.